### ISSUE TITLE
How to use it with a different SHM

### ISSUE THREAD

**Issue by amrx101** [mheese/rust-pkcs11#21]
Hi,

I have a [HSM](https://github.com/MicrochipTech/cryptoauthlib/wiki/PKCS11-Linux-Setup) which can be set up a PKCS11 provider. I was wondering would this library run out of box with the provided PKCS11(after configuring pkcs11 location) or would we need some modifications?

Any pointers would be helpful.

Thanks

**Comment by mheese** [comment]
**TL;DR:** yes, just point to that module and you are good to go.

That is the general idea of course. SoftHSM is only used as a testing harness as it becomes very impractical/impossible for me to test against different hardware.

The team around SoftHSM has done a very good job to implement the PKCS#11 standard according to the required business logic and it goes far beyond simple mocking. So it is ideal to test this rust library.

In my experience developing in general against SoftHSM first gives you confidence into your own HSM dependant code, however, moving to a "real" HSM usually always requires additional considerations that one often does not think about. Nonetheless it is a good starting point.

Also, the behaviour between different HSMs is often very different and also depends on the configuration of the HSM itself.

To complicate things even further: not al HSMs have very "clean" implementations of pkcs11 themselves, so things that work and should work with one device, are not guaranteed to work on another and vice versa.

Does that answer your question?

**Comment by mheese** [comment]
btw, if you are successful in using this HSM with this library, please shoot me a note. I'd like to start to keep a compatibility list.

**Comment by woodruffw** [comment]
JFYI: I'm experimenting with the library on the [Nitrokey HSM 2](https://www.nitrokey.com), which uses a [CardConnect SmartCard-HSM](https://www.smartcard-hsm.com/) internally.

I'll let you know if I run into any problems (and should be able to fix and upstream many of them myself 🙂).

**Comment by mheese** [comment]
Closing this now. Please reopen if this did not answer your question.
